tony-core/src/main/java/com/linkedin/tony/TaskMonitor.java

```diff
@@ -51,15 +92,79 @@ public void run() {
   }
 
   private void refreshMetrics() {
-    resourceCalculator.updateProcessTree();
-    refreshMaxMemoryBytes();
+    refreshMemoryBytesMetrics();
+    if (isGpuMachine) {
+      refreshGPUMetrics();
+    }
+    numRefreshes++;
   }
 
-  private void refreshMaxMemoryBytes() {
+  private void refreshMemoryBytesMetrics() {
+    resourceCalculator.updateProcessTree();
     double memoryBytes = resourceCalculator.getRssMemorySize();
-    if (memoryBytes > maxMemoryBytes.getValue()) {
-      maxMemoryBytes.setValue(memoryBytes);
-      metrics.setMetric(MAX_MEMORY_BYTES_INDEX, maxMemoryBytes);
+    setMaxMetrics(MAX_MEMORY_BYTES_INDEX, memoryBytes);
+    setAvgMetrics(AVG_MEMORY_BYTES_INDEX, memoryBytes);
+  }
+
+  private void refreshGPUMetrics() {
+    try {
+      GpuDeviceInformation gpuInfo = gpuDiscoverer.getGpuDeviceInformation();
+
+      double maxGpuUtilization = gpuInfo.getGpus().stream()
+          .mapToDouble(x -> x.getGpuUtilizations().getOverallGpuUtilization())
+          .max()
+          .getAsDouble();
+      double avgGpuUtilization = gpuInfo.getGpus().stream()
+          .mapToDouble(x -> x.getGpuUtilizations().getOverallGpuUtilization())
+          .average()
+          .getAsDouble();
+      double maxFBMemoryUsage = gpuInfo.getGpus().stream()
+          .mapToDouble((x ->
+              ((double) x.getGpuMemoryUsage().getUsedMemoryMiB() / x.getGpuMemoryUsage().getTotalMemoryMiB())))
+          .max()
+          .getAsDouble();
+      double avgFBMemoryUsage = gpuInfo.getGpus().stream()
+          .mapToDouble((x ->
+              ((double) x.getGpuMemoryUsage().getUsedMemoryMiB() / x.getGpuMemoryUsage().getTotalMemoryMiB())))
+          .average()
+          .getAsDouble();
+
+      setMaxMetrics(MAX_GPU_UTILIZATION_INDEX, maxGpuUtilization);
+      setAvgMetrics(AVG_GPU_UTILIZATION_INDEX, avgGpuUtilization);
+      setMaxMetrics(MAX_FB_MEMORY_USAGE_INDEX, maxFBMemoryUsage);
+      setAvgMetrics(AVG_FB_MEMORY_USAGE_INDEX, avgFBMemoryUsage);
+    } catch (Exception e) {
+      // Follow YARN's GPUDiscoverer mechanism of capping number of gpu metrics query
+      if (gpuDiscoverer.getNumOfErrorExecutionSinceLastSucceed() >= Constants.MAX_REPEATED_GPU_ERROR_ALLOWED) {
+        String msg = "Failed to collect GPU metrics, this is not a GPU machine";
```

Comment:
`this is not a GPU machine` is confusing, remove it. it could be something wrong with GPU driver.